library/x509_create.c

```diff
@@ -125,81 +125,296 @@ static const x509_attr_descriptor_t *x509_attr_descr_from_name( const char *name
     char *d = data;
 
-    /* Clear existing chain if present */
-    mbedtls_asn1_free_named_data_list( head );
+    /*
+     * As defined by RFC 1779, the DN is constructed from
+     * <string> | <key> <optional-space> "=" <optional-space> <string>
+     * where optional space is
+     * ( <CR> ) *( " " )
+     *
+     * When parsing the value:
+     * 1. Skip whitespaces.
+     * 2. Check if it is a quoted string.
+     * 3. Read the characters, to evaluate as a valid value.
+     * 4. Skip whitespaces until reaching separator character (',').
+     */
+
+    /*
+     * Skip whitespaces
+     */
+    while( *cur < end && *( *cur ) == ' ' )
+    {
+        ( *cur )++;
+    }
 
-    while( c <= end )
+    if( *cur == end || *( *cur ) == ',' )
     {
-        if( in_tag && *c == '=' )
-        {
-            if( ( attr_descr = x509_attr_descr_from_name( s, c - s ) ) == NULL )
-            {
-                ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
-                goto exit;
-            }
+        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+        goto exit;
+    }
+    /*
+     * Check if it is a quoted string
+     */
+    else if( *( *cur ) == '\"' )
+    {
+        in_quot = 1;
+        ( *cur )++;
+    }
 
-            oid = attr_descr->oid;
-            s = c + 1;
-            in_tag = 0;
-            d = data;
+    while( *cur < end )
+    {
+        /*
+         * Check whether reached end of value to evaluate,
+         * whether it is a closing quotation, or a separator character
+         */
+        if( in_quot == 1 && *( *cur ) == '\"' )
+        {
+            /*
+             * Increase the current position to after the closing quotation
+             */
+            ( *cur )++;
+            /*
+             * Mark that quotation has ended.
+             */
+            in_quot = 0;
+            break;
         }
-
-        if( !in_tag && *c == '\\' && c != end )
+        else if( in_quot == 0 && *( *cur ) == ',')
         {
-            c++;
+            break;
+        }
 
-            /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+        if( *cur != end && *( *cur ) == '\\' )
+        {
+            ( *cur )++;
+            ret = x509_dn_check_special_chars( cur, end );
+            if( ret != 0 )
             {
                 ret = MBEDTLS_ERR_X509_INVALID_NAME;
                 goto exit;
             }
         }
-        else if( !in_tag && ( *c == ',' || c == end ) )
+
+        /*
+         * Fill the current data pointer with the unescaped character.
+         */
+        if( d - data < MBEDTLS_X509_MAX_DN_NAME_SIZE )
+            *( d++ ) = *( *cur );
+        /*
+         * A whitespace can be part of the string,
+         * but it can also be a trailing whitespace, so don't set the
+         * value length until the character is  not a whitespace
+         */
+        if( *( *cur ) != ' ' )
+            value_len = d - data;
+
+        if( value_len >= MBEDTLS_X509_MAX_DN_NAME_SIZE )
```

Comment:
I'm sorry, but I don't understand what you mean

Reply:
I think in the first paragraph, I was describing a situation that would cause a failure even though there is enough space in the data buffer to hold the input. (Note that A and B are simple characters). Also, it would be nice to have a test case for this, which I do not think we currently have.

I think in the second paragraph I am pointing out that the condition `value_len > MBEDTLS_X509_MAX_DN_NAME_SIZE` will never be true. (Note that the == is not there on purpose) It is dead code because `value_len = d - data` in line 321, but `d - data` can never be greater than `MBEDTLS_X509_MAX_DN_NAME_SIZE` because of the check inline 313.

In summary, I am suggesting that we should rearrange a bit the code here to make sure that these strange corner cases do not happen and also to make the code easier to read. For example, replacing the lines:
```c
        /*
         * Fill the current data pointer with the unescaped character.
         */
        if( d - data < MBEDTLS_X509_MAX_DN_NAME_SIZE )
            *( d++ ) = *( *cur );
        /*
         * A whitespace can be part of the string,
         * but it can also be a trailing whitespace, so don't set the
         * value length until the character is  not a whitespace
         */
        if( *( *cur ) != ' ' )
            value_len = d - data;

        if( value_len >= MBEDTLS_X509_MAX_DN_NAME_SIZE )
        {
            ret = MBEDTLS_ERR_X509_INVALID_NAME;
            goto exit;
        }
```
with something like:
```c
        /*
         * Fill the current data pointer with the unescaped character.
         */
        if( d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE && *( *cur ) != ' ' )
        {
            /* The data buffer is not large enough to hold the value */
            ret = MBEDTLS_ERR_X509_INVALID_NAME;
            goto exit;
        }
        else if( d - data < MBEDTLS_X509_MAX_DN_NAME_SIZE )
        {
            /*
             * Assume that the current character is part of the value tag, but
             * this might actually not be the case if it is just trailing whitespace
             */
            *( d++ ) = *( *cur );
        }

        /*
         * A whitespace can be part of the string,
         * but it can also be a trailing whitespace, so don't set the
         * value length until the character is  not a whitespace
         */        
        if ( *( *cur ) != ' ' )
            value_len = d - data;
```

Reply:
But the first paragraph is actually a failure, because  in this case length of `A<lots of whitespaces>B` is `1 (A) + BEDTLS_X509_MAX_DN_NAME_SIZE -1 ( whitespaces) + 1 ( B) = BEDTLS_X509_MAX_DN_NAME_SIZE + 1`
There is test case for this [here](https://github.com/ARMmbed/mbedtls/pull/784/files#diff-0807f152469bce0d302ec7282849e58bR170)

As for the second paragraph, I believe I meant the following:
` d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE - 1` -> `d++` -> ` d - data == MBEDTLS_X509_MAX_DN_NAME_SIZE`
After that, if current character is not white space, we now have `value_len == MBEDTLS_X509_MAX_DN_NAME_SIZE` which should cause an error I believe.

Anyway, I like your suggestion better, as it's clearer

Reply:
There is minor issue with your suggestion:
`sizeof(data) == MBEDTLS_X509_MAX_DN_NAME_SIZE ` but the longest input string can be 255, to have the terminating null character.
I'll change your suggestion a bit